### PR TITLE
Add the possibility to round only specific corners.

### PR DESCRIPTION
**Pull request by josketres**
This changeset adds the possibiltiy to specify which corners should be rounded. (see screenshot)
This is only implemented for RoundedDrawable.
I will add it to RoundedImageView in a further pull request.

### CODE REVIEW

**Comment by vinc3m1** [comment]
Nice! Can you turn on overdraw in dev settings and see what this looks like? I'm concerned about performance.

I saw another lib out there using a method that uses custom paths, but i'm not sure about the performance implications of `drawPath` vs `drawRoundRect` (hw accel, antialiasing, etc).

Also we might be able to minimize drawn area by limiting it to the rounded radius instead of quarter of a drawable.

**Comment by josketres** [comment]
Good idea, I'll reduce the drawn area by limiting it to the rounded radius!

**Comment by josketres** [comment]
There's still an overdraw, but it's limited to the corners. See screenshot: screenshot-overdraw-debug

File: roundedimageview/src/main/java/com/makeramen/roundedimageview/RoundedDrawable.java [hunk 1]
```diff
@@ -278,14 +283,78 @@ public void draw(Canvas canvas) {
     } else {
       if (mBorderWidth > 0) {
         canvas.drawRoundRect(mDrawableRect, Math.max(mCornerRadius, 0),
-            Math.max(mCornerRadius, 0), mBitmapPaint);
+                Math.max(mCornerRadius, 0), mBitmapPaint);
         canvas.drawRoundRect(mBorderRect, mCornerRadius, mCornerRadius, mBorderPaint);
+        redrawBitmapForSquareCorners(canvas);
+        redrawBorderForSquareCorners(canvas);
```
Comment (vinc3m1) [inline roundedimageview/src/main/java/com/makeramen/roundedimageview/RoundedDrawable.java]: why twice?

**Comment by vinc3m1** [comment]
Hmmm... how hard would it be to allow each corner to have a different radius? What if we draw each quarter of the bitmap separately once (instead of overdrawing), based on the radius set for that corner? You think that'd be feasible?

**Comment by josketres** [comment]
Drawing each quarter separately is not an option because you still have to solve the problem of drawing rectangles with just one rounded corner. Otherwise the center of the image would be not fully drawn.

This implementation is based on redrawing the square corners over a roundrectangle (`Canvas.drawRoundRect`). In order to allow each corner to have different radius we need to change the implementation to one based on custom paths (`Canvas.drawPath`)... I think such a solution needs a lot more of work.

**Comment by josketres** [comment]
After watching the nice video of Romain Guy (https://www.parleys.com/tutorial/filthy-rich-android-clients-2) I realized that I was instanciating RectF objects on every paint. So i fixed this. See 6c6882e0d89f0eb676c5f95b984aac469beab802

**Comment by vinc3m1** [comment]
I'm still concerned about this making the API too complex. I'd rather have an API that supports different rounds per corner if we'll eventually go that way than picking which corners to round right now and re-adjusting the api again later.

I'd be ok with throwing an exception for now if the corner radii aren't all the same or 0 and saying different radii aren't supported yet if that makes sense?

**Comment by josketres** [comment]
I agree! Supporting different radius per corner is a more elegant API.
So I'll change that part and throw an IllegalArgument or NotImplemented exception for now if not all radius are the same or 0.

**Comment by josketres** [comment]
I'll let `riv_corner_radius` as a shortcut attribute to set all the corners at once.

**Comment by josketres** [comment]
Introducing this change would break the existing API. 
For example, what should we do with `RoundedImageView.getCornerRadius`?

**Comment by vinc3m1** [comment]
Hm, we could just have that return the max radius? And make sure it's in javadoc.

**Comment by josketres** [comment]
Done, I've changed the API to support setting corner radius per corner.
Please take a look at it...
The changeset is getting complexer =(

**Comment by vinc3m1** [comment]
Sorry haven't had much time to look at this lately, but I did just notice drawRoundRect got support for different corner radii in API 21+:
[http://developer.android.com/reference/android/graphics/Canvas.html#drawRoundRect(float, float, float, float, float, float, android.graphics.Paint)](http://developer.android.com/reference/android/graphics/Canvas.html#drawRoundRect(float, float, float, float, float, float, android.graphics.Paint))

**Comment by gabrielmoreira** [comment]
I need some way to disable bottom corners, preserving upper corners, and all corners using same radius.

Something like riv_corner_radius="4dp", riv_corners="top_left|top_right" 
where riv_corners can be [top_left, top_right, bottom_left, bottom_right] 

There are another API 14+ alternative?

**Comment by josketres** [comment]
This weekend I'll try to use the new version of drawRoundRect for API 21+ and let the actual implementation (which has some overdraw) for API < 21

**Comment by josketres** [comment]
@vinc3m1  the method that you mention is not providing support for different corner radii, it's just an utility method that accepts 4 float values instead of 1 RectF.

http://developer.android.com/reference/android/graphics/Canvas.html#drawRoundRect(float, float, float, float, float, float, android.graphics.Paint)

So, please take a look at the changeset and let's decide if it's OK to merge the pull request. I'm already working with this implementation in my app. It would be nice if we can integrate it in this lib.

**Comment by vinc3m1** [comment]
@josketres good call I see that now. Will take a look today.

**Pull request merged**